src/internal_modules/roc_core/string_list.cpp

```diff
@@ -54,14 +59,51 @@ const char* StringList::nextof(const char* str) const {
         roc_panic("stringlist: string doesn't belong to the list");
     }
 
-    const char* ptr = str + strlen(str) + 1;
+    const Header* str_header =
+        (const Header*)(ROC_CONTAINER_OF(const_cast<char*>(str), Header, str_));
+    const size_t align = AlignOps::align_as(sizeof(Header) + str_header->len + 1, 8);
+    const char* ptr = (const char*)str_header + align + sizeof(Footer);
     roc_panic_if(ptr > end);
```

Comment:
A few ideas how to cleanup nextof() and prevof()

Reply:
1. To avoid const casts in user code, we can fix definition of ROC_CONTAINER_OF as follows:

```cpp
#define ROC_CONTAINER_OF(ptr, type, member)                                              \
    (reinterpret_cast<type*>(                                                            \
        const_cast<char*>((const char*)(ptr)-offsetof(type, member))))
```

Then we can use it like this:

```cpp
const Header* str_header = ROC_CONTAINER_OF(str, const Header, str);
```

Reply:
2. Let's extract this piece:

```cpp
    const char* begin = &data_[0];
    const char* end = &data_[0] + data_.size();
    // check_pointer_(str);

    if (str < begin || str >= end) {
        roc_panic("stringlist: string doesn't belong to the list");
    }
```

into a private method, say, `check_is_member_(str)` (we have similar in core::List).

Reply:
3. I played with the code a bit, and it appears to me that the code will be much simpler if `header->len` and `footer->len` would contain the length of the whole block (header + string + alignment + footer). What do you think?

Here is an example of how nextof() and prevof() could look like:

```cpp
    const Header* str_header = ROC_CONTAINER_OF(str, const Header, str);
    if (str_header == back_) {
        return NULL;
    }

    const Header* next_header =
        (const Header*)((const char*)str_header + str_header->len);

    return next_header->str;
```

```cpp
    const Header* str_header = ROC_CONTAINER_OF(str, const Header, str);
    if (str_header == front_) {
        return NULL;
    }

    const Footer* prev_footer = (const Footer*)((const char*)str_header - sizeof(Footer));
    const Header* prev_header =
        (const Header*)((const char*)str_header - prev_footer->len);

    return prev_header->str;
```

(I didn't test it)

Reply:
push_unique was the only reason why i had the keep the length to the actual string otherwise i couldnt get the string length to do the comparison in there, but i could remove it if you’d like me to

Reply:
I wanted to remove push_unique() and instead add `bool find(const char*)`, and use:

```cpp
if (!list.find("foo")) {
    list.push_back("foo");
}
```

IIRC there were only 2 usages of push_unique actually.

You can do this if you wish, or you can keep push_unique but don't bother making it efficient/clean, and I'll remove it later. In both ways I guess you can store whole block size in `len`.

Also I think `push_back(begin, end)` overload may be removed for simplicity, as it's not used IIRC.

Reply:
ill replace push_unique() with `find()`
as for `push_back(begin, end)` one usage i could possibly think of is maybe wanting push a string with a null character within it, like if you ran it through `strtok` for example, but im find with removing it if you still think it wont be useful

Reply:
Thanks. Makes sense, we indeed have non-null-terminated strings inside some ragel parsers, push_back(begin, end) can be useful here, let's keep it.